DirectProgramming/DPC++FPGA/include/streaming_cholesky.hpp

```diff
@@ -0,0 +1,248 @@
+#ifndef __STREAMING_CHOLESKY_HPP__
+#define __STREAMING_CHOLESKY_HPP__
+
+#include "constexpr_math.hpp"
+#include "tuple.hpp"
+#include "unrolled_loop.hpp"
+
+namespace fpga_linalg {
+
+/*
+  Cholesky decomposition - Computes L such that A=LL* where:
+  - A is the input matrix (hermitian, positive definite)
+  - L is a lower triangular matrix
+  - L* is the conjugate transpose of L
+
+  This function implements a modified version of the Cholesky–Banachiewicz
+  algorithm.
+  Pseudo code:
+
+  int row_size = 0;
+  for (column = 0; column <= row_size; column++) {
+    for (row = column; row < rows; row++) {
+      float sum = 0;
+      for (k = 0; k < column; k++)
+        sum += L[row][k] * L[column][k];
+
+      if (row == column)
+        L[row][column] = sqrt(A[row][row] - sum);
+      else
+        L[row][column] = (A[row][column] - sum) / L[column][column];
+    }
+  }
+
+  The input and output matrices are consumed/produced from/to pipes.
+*/
+template <typename T,       // The datatype for the computation
+          bool is_complex,  // True if T is ac_complex<X>
+          int rows,         // Number of rows==columns in the A matrices
+          int raw_latency,  // Read after write latency (in iterations) of
+                            // the triangular loop of this function.
+                            // This value depends on the FPGA target, the
+                            // datatype, the target frequency, etc.
+                            // This value will have to be tuned for optimal
+                            // performance. Refer to the Triangular Loop
+                            // design pattern tutorial.
+                            // In general, find a high value for which the
+                            // compiler is able to achieve an II of 1 and
+                            // go down from there.
+          int pipe_size,    // Number of elements read/write per pipe operation
+                            // to read the input matrix
```

Comment:
some static asserts on these would be useful to make sure what the user passed makes sense.